api-server/stack-test-suite/tests/in_memory.rs

```diff
@@ -0,0 +1,1160 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use api_server_common::storage::{
+    impls::in_memory::transactional::TransactionalApiServerInMemoryStorage,
+    storage_api::{ApiServerStorageWrite, ApiServerTransactionRw, Transactional},
+};
+use blockchain_scanner_lib::{
+    blockchain_state::BlockchainState, sync::local_state::LocalBlockchainState,
+};
+use chainstate_test_framework::TestFramework;
+use common::{
+    chain::config::create_unit_test_config,
+    primitives::{BlockHeight, Idable},
+};
+use hex::ToHex;
+use rstest::rstest;
+use serde_json::json;
+use std::{net::TcpListener, sync::Arc};
+use test_utils::random::{make_seedable_rng, Rng, Seed};
+use web_server::{api::web_server, APIServerWebServerState};
+
+#[tokio::test]
+async fn server_status() {
+    let url = "/";
+
+    let listener = TcpListener::bind("127.0.0.1:0").unwrap();
+    let socket = listener.local_addr().unwrap();
+
+    tokio::spawn(async move {
+        let web_server_state = {
+            let chain_config = Arc::new(create_unit_test_config());
+            let storage = TransactionalApiServerInMemoryStorage::new(&chain_config);
+
+            APIServerWebServerState {
+                db: Arc::new(storage),
+                chain_config: Arc::clone(&chain_config),
+            }
+        };
+
+        web_server(listener, web_server_state).await
+    });
+
+    let response = reqwest::get(format!("http://{}:{}{url}", socket.ip(), socket.port()))
+        .await
+        .unwrap();
+
+    assert_eq!(response.status(), 200);
+    assert_eq!(response.text().await.unwrap(), r#"{"versions":["1.0.0"]}"#);
+}
+
+#[tokio::test]
+async fn bad_request() {
+    let url = "/non-existent-url";
+
+    let listener = TcpListener::bind("127.0.0.1:0").unwrap();
+    let socket = listener.local_addr().unwrap();
+
+    tokio::spawn(async move {
+        let web_server_state = {
+            let chain_config = Arc::new(create_unit_test_config());
+            let storage = TransactionalApiServerInMemoryStorage::new(&chain_config);
+
+            APIServerWebServerState {
+                db: Arc::new(storage),
+                chain_config: Arc::clone(&chain_config),
+            }
+        };
+
+        web_server(listener, web_server_state).await
+    });
+
+    let response = reqwest::get(format!("http://{}:{}{url}", socket.ip(), socket.port()))
+        .await
+        .unwrap();
+
+    assert_eq!(response.status(), 400);
+    assert_eq!(response.text().await.unwrap(), r#"{"error":"Bad request"}"#);
+}
+
+mod v1_block {
+    use super::*;
+
+    #[tokio::test]
+    async fn invalid_block_id() {
+        let url = "/api/v1/block/invalid-block-id";
+
+        let listener = TcpListener::bind("127.0.0.1:0").unwrap();
+        let socket = listener.local_addr().unwrap();
+
+        tokio::spawn({
+            async move {
+                let web_server_state = {
```

Comment:
In addition to the race @TheQuantumPhysicist pointed out, also note that the tokio task is not cleaned up. Not sure to what extent it matters though.

Also a tiny style comment: There seems to be a redundant set of braces around the `async` block. Is that inserted by cargo fmt?

Reply:
Sorry that was me - a left over from when I was localizing a binding but have deleted them without cleaning up